src/transformStateWithClones.js

```diff
@@ -1,13 +1,33 @@
-'use strict';
-
-/**
- * @param {Object} state
- * @param {Object[]} actions
- *
- * @return {Object[]}
- */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = { ...currentState };
```

Comment:
There's no need to create a new copy of currentState in the 'removeProperties' case since you are not mutating the original state object but rather modifying the current copy which is the desired behavior according to the checklist.